### PR TITLE
batcheval: use write-batches to apply small SSTables

### PR DESCRIPTION
**Pull request by dt**
Adding very small SSTables is not great: the benefits of adding the raw SST, in terms of avoiding per-key overhead, are small for small SSTs, while the fixed costs of adding a file are the same — triggering flushes, adding to number of files that need to be compacted, etc — so when files become too small, they can actually be more expensive, per key, than just writing their contents via the regular write path.

This patch gives the caller control of how SSTs are ingested — either via the usual direct AddFile or by constructing a normal write-batch instead.

Requests marked for write-based ingestion can additionally skip the back-pressure mechanisms added for SSTable additions — since we’re not creating files directly, the normal write back pressure still applies, and these requests also don’t risk of bloating the file count limits with tiny files.

Release note: none.

### CODE REVIEW

**Review by ajwerner — APPROVED** [review]
:lgtm:

> Reviewed 12 of 12 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41705#-:-LrR9w_j9HuSeG-EF6Uk:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[pkg/storage/bulk/sst_batcher.go, line 402 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41705#-LrR9Qw47XWe2L2kKTx_:-LrR9Qw47XWe2L2kKTxa:b-1jn86q) ([raw file](https://github.com/cockroachdb/cockroach/blob/77a905ee899ae4628da7e0fa9c9a14044db173be/pkg/storage/bulk/sst_batcher.go#L402)):*
it this

**Review by dt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41705)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner and @nvanbenschoten)

___
*[pkg/storage/bulk/sst_batcher.go, line 402 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41705#-LrR9Qw47XWe2L2kKTx_:-LrTt3dh3idE3xmeDUkq:b3l2qdo) ([raw file](https://github.com/cockroachdb/cockroach/blob/77a905ee899ae4628da7e0fa9c9a14044db173be/pkg/storage/bulk/sst_batcher.go#L402)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

it this
</blockquote></details>

Done (re-worded this comment as it was still written assuming it was in cmd_add_sstable) too.

File: pkg/roachpb/api.proto [hunk 1]
```diff
@@ -1471,6 +1471,8 @@ message AddSSTableRequest {
   // MVCCStats, instead of computing the stats for the SSTable by iterating it.
   // Including these stats can make the evaluation of AddSSTable much cheaper.
   storage.engine.enginepb.MVCCStats mvcc_stats = 4 [(gogoproto.customname) = "MVCCStats"];
+
+  bool ingest_as_writes = 5;
```
Comment (nvanbenschoten) [inline pkg/roachpb/api.proto]: Could you give this a comment?

File: pkg/storage/batcheval/cmd_add_sstable.go [hunk 2]
```diff
@@ -174,6 +174,24 @@ func EvalAddSSTable(
 	stats.ContainsEstimates = !args.DisallowShadowing
 	ms.Add(stats)
 
+	if args.IngestAsWrites {
+		log.VEventf(ctx, 2, "ingesting SST (%d keys/%d bytes)via regular write batch", stats.KeyCount, len(args.Data))
```
Comment (nvanbenschoten) [inline pkg/storage/batcheval/cmd_add_sstable.go]: Missing space before "via"

**Review by nvb — APPROVED** [review]
LGTM, though this seems like the kind of thing we'd want a test for.

**Review by dt — COMMENTED** [review]
Thanks, added a test that confirms we don't see the SST application when using the flag.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41705#-:-LrUYQAC8yY-KFjGP52h:b-t34it2)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner and @nvanbenschoten)

___
*[pkg/roachpb/api.proto, line 1475 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41705#-LrQH_h0q4iAi24Las4S-r2-1475:-LrUYOU-6FapTceXRqrC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb7902bcc8ee841a9b8f925ce442c5b2b94fe02b/pkg/roachpb/api.proto#L1475)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Could you give this a comment?
</blockquote></details>

Done.
___
*[pkg/storage/batcheval/cmd_add_sstable.go, line 178 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41705#-LrQH_h0q4iAi24Las4V-r2-178:-LrUPxyc1uE5M0hRnqT4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb7902bcc8ee841a9b8f925ce442c5b2b94fe02b/pkg/storage/batcheval/cmd_add_sstable.go#L178)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Missing space before "via"
</blockquote></details>

Done.

**Review by petermattis — COMMENTED** [review]
It rubs me slightly wrong to see a parameter on `AddSSTable` which indicates that an sstable should be converted to a batch when the request is evaluated. I wonder if we should perform the size check within the `AddSSTable` command itself rather than adding a parameter to the command. Certainly, it is something of a foot-gun to specify `ingestAsSWrite = true` with a very large sstable.

I wonder if we could push down the conversion from sstable to batch even lower in the system. Could sstable-based snapshots benefit from this optimization as well? I haven't looked at that code in a while, so @ajwerner or @nvanbenschoten will have to chime in.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41705#-:-LrZ1qXR4nRr89153Tfq:b-2rhvui)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner, @dt, and @nvanbenschoten)

___
*[pkg/storage/batcheval/cmd_add_sstable.go, line 187 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/41705#-LrZ0BrE2f0Ale1mRaaj:-LrZ0BrE2f0Ale1mRaak:b-1qrg04) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6263d6ac079a0b0872ea4a83030363582fed118/pkg/storage/batcheval/cmd_add_sstable.go#L187)):*
> ```Go
> 				break
> 			}
> 			if err := batch.Put(dataIter.UnsafeKey(), dataIter.UnsafeValue()); err != nil {
> ```

This is correct for import/restore, but not for rebalancing as you're converting any record in the sstable into a `Put`. For rebalancing, we can have `Delete` and `DeleteRange` operations. I suspect that we should pull this functionality out into a function in `storage/engine` that can be tested in isolation. Happy to have someone on the storage team do this if you'd prefer.
___
*[pkg/storage/bulk/sst_batcher.go, line 37 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/41705#-LrZ11fbCOd5SBWjuSVB:-LrZ11fbCOd5SBWjuSVC:b7tpsej) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6263d6ac079a0b0872ea4a83030363582fed118/pkg/storage/bulk/sst_batcher.go#L37)):*
> ```Go
> 		"kv.bulk_io_write.small_write_size",
> 		"size below which a 'bulk' write will be performed as a normal write instead",
> 		1<<18,
> ```

Nit 1: can you add a comment about what this constant translates to?
Nit 2: I think this is easier to read as `256 << 10` as `<< 10` is more clearly `* 1024` (at least it is to me).

**Review by dt — COMMENTED** [review]
> I wonder if we should perform the size check within the AddSSTable command itself rather than adding a parameter to the command.

I actually had the size check in eval initially, but switched it to this client-provided flag for two reasons: 
 a) so that we could tell what the request is going to do (i.e. add a file or not) when it gets to eval, so that we can decide if we need to rate limit it based on file counts and b) it made it much easier to test.

> Certainly, it is something of a foot-gun to specify ingestAsWrite = true with a very large sstable.

Fair. We could add a size check and reject requests to add as writes on >10mb or something? This RPC is already a pretty direct, no-guardrails method to interact directly with storage -- you can already e.g. break MVCC or transactions if you misuse it -- but that's mostly by design as its behavior is tailored to exactly what the bulk-ingestion pipeline that is built on top of it needs.

> I wonder if we could push down the conversion from sstable to batch even lower in the system.

I'm not sure about pushing the conversion below eval -- if we put it in raft as `AddSSTable`, we're going to do the sideload file IO, the current application logic is pretty hard-coded, and I'm just generally wary of pushing anything too fancy under raft?

Side note: I realize the name vs possible behavior conflict a bit after this change, and I'm inclined to eventually rename this method / request from `AddSSTable` to `Ingest`, since it both does more than just add sstables and, potentially, doesn't even always add sstables, but for now want to keep these changes as small as possible until we figure out what what will/will not need to be backported to resolve the sluggish IMPORTs a customer currently testing the 19.2 beta was seeing.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41705#-:-LrZ3-jr-T87whehyxEf:b-fauxmx)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner, @dt, and @nvanbenschoten)

**Review by petermattis — COMMENTED** [review]
> I actually had the size check in eval initially, but switched it to this client-provided flag for two reasons:
a) so that we could tell what the request is going to do (i.e. add a file or not) when it gets to eval, so that we can decide if we need to rate limit it based on file counts and b) it made it much easier to test.

Ack. I'm sensitive to these arguments, but also want to be careful about adding too many bandaids on top of bandaids. Sstable ingestion is heavyweight for a small sstable. No question. Yet we could also make ingestion faster in Pebble. See https://github.com/cockroachdb/pebble/issues/25 for one idea. I can also imagine performing this transformation from sstable to write batch inside `pebble/DB.Ingest`. There are other bits that would have to be fixed in order to do this, such as figuring out how to properly backpressure ingestion. And any improvement here likely wouldn't happen for 20.1 which doesn't mesh with your desire to get some near term improvements in.

I'm not going to stand in the way of this PR, just want to make sure we're not painting ourselves into a corner.

> Fair. We could add a size check and reject requests to add as writes on >10mb or something? This RPC is already a pretty direct, no-guardrails method to interact directly with storage -- you can already e.g. break MVCC or transactions if you misuse it -- but that's mostly by design as its behavior is tailored to exactly what the bulk-ingestion pipeline that is built on top of it needs.

Good point about the existing possibility for misuse. 

> I'm not sure about pushing the conversion below eval -- if we put it in raft as AddSSTable, we're going to do the sideload file IO, the current application logic is pretty hard-coded, and I'm just generally wary of pushing anything too fancy under raft?

I generally concur, yet this is one of the few places that such logic is ok as we're deciding between two alternatives which should have no functional difference on the KV storage state. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41705#-:-Lr_lCBFCC-8Y5qji5GY:b-fhmyv2)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner, @dt, and @nvanbenschoten)

**Comment by dt** [comment]
Yeah, I should have scoped my reservations about pushing this down to pushing it down _inside kv_, but in the future I could absolutely could see the storage engine doing smarter things when handed a file, like buffering instead of ingesting if it is too small, splitting if it that spanned existing data but could otherwise ingest lower, etc. 

I'm not too worried about the flag painting us into corners with this change -- if/when we have a new, smarter low level ingestion behavior, we could simply ignore this flag if we knew our new behavior was strictly better, or version-gate switching over.

**Review by dt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41705)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner, @nvanbenschoten, and @petermattis)

___
*[pkg/storage/batcheval/cmd_add_sstable.go, line 187 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/41705#-LrZ0BrE2f0Ale1mRaaj:-LrdTqH9EQqnigNbxPXB:b-39owc3) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6263d6ac079a0b0872ea4a83030363582fed118/pkg/storage/batcheval/cmd_add_sstable.go#L187)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

This is correct for import/restore, but not for rebalancing as you're converting any record in the sstable into a `Put`. For rebalancing, we can have `Delete` and `DeleteRange` operations. I suspect that we should pull this functionality out into a function in `storage/engine` that can be tested in isolation. Happy to have someone on the storage team do this if you'd prefer.
</blockquote></details>

Yep, I should have left a comment (and did now): this simple iteration+Put is OK here specifically because we're in this RPC, which is only use in bulk-ingest (import/restore), and already assumes the SST is all Sets in other places.

Called this out in a comment.
___
*[pkg/storage/bulk/sst_batcher.go, line 37 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/41705#-LrZ11fbCOd5SBWjuSVB:-LrdVKa18nvCdVfaZMML:bacuxac) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6263d6ac079a0b0872ea4a83030363582fed118/pkg/storage/bulk/sst_batcher.go#L37)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Nit 1: can you add a comment about what this constant translates to?
Nit 2: I think this is easier to read as `256 << 10` as `<< 10` is more clearly `* 1024` (at least it is to me).
</blockquote></details>

Done. 
(also bumped it to 400kb, since I was still seeing lots of l0 files and not *that* many flushes at 256).

**Comment by dt** [comment]
@petermattis Do you still have reservations or concerns on this one?

**Comment by dt** [comment]
Added a TODO pointing to that comment.

Offline convo with @ajwerner: I still think this is the most promising fix we can get into 19.2, while we work on something better for 20.1, as outlined in that issue comment, and I think he agrees as mentioned above, but I would ideally still like to go/no-go calls from @nvanbenschoten and @petermattis before back-porting it though.

**Comment by dt** [comment]
TFTRs everyone!

bors r=ajwerner,nvanbenschoten

**Pull request merged**